### PR TITLE
Doc: Set up Breathe

### PR DESCRIPTION
**Pull request by reneme**
This adds [breathe](https://github.com/breathe-doc/breathe) as a bridge between Doxygen and Sphinx. It should allow us to replace _a lot_ of duplicated information simply with pointers to the information from Doxygen comments.

Also, I added a (super simplistic) stand-in if breathe is disabled or not available during `./configure.py`. Basically, it just writes something along these lines into the documentation:

```
[Missing Breathe Content: Botan::TLS::Callbacks]
```

No adaptions to the reStructuredText documentation have been done, yet. That will come in later PRs. If you want to test the integration locally, just go to `hash.rst` and replace the entire `.. cpp:class:: HashFunction` block with:

```
.. doxygenclass:: Botan::HashFunction
   :members:
```

... more fine tuning might be needed to further filter the displayed information, but it should show that it works.

### CODE REVIEW

**Comment by reneme** [comment]
Note that this now also installs `sphinx-togglebutton` (rationale: #3470)

**Review by randombit — COMMENTED** [review]
Since lacking Breathe results in somewhat substandard documentation, the usage should be noted in the manual, changelog, and also the migration guide. I expect even with all that the initial packaging for distributions will miss the new dependency.

File: doc/migration_guide.rst [hunk 1]
```diff
@@ -7,6 +7,16 @@ This guide attempts to be, but is not, complete. If you run into a problem while
...
 
+Documentation
+-------------
+
+Most of the API references are now powered by `Breathe
+<https://github.com/breathe-doc/breathe>`_. Without Breathe the documentation
+will lack almost all API details and will be much less helpful. Additionally,
+the Sphinx extension `togglebutton
+<https://pypi.org/project/sphinx-togglebutton/>`_ must be installed to
+successfully build the documentation.
```
Comment (randombit) [inline doc/migration_guide.rst:18]: Woah - is this really correct? We *must* have this extension, to build the docs at all? I don't think that can work.

Per https://repology.org/project/python:sphinx-togglebutton/versions effectively zero Linux distributions include this package, which would mean that everyone installing the Linux packages from their distro would get no (or broken) documentation.

**Comment by randombit** [comment]
I think there are some major unresolved issues here and since our adopting vs not adopting this is not a SemVer issue, I propose postponing to after 3.0 is released.

File: doc/migration_guide.rst (hunk 1 above)
Comment (reneme) [reply doc/migration_guide.rst:18]: Mhm, indeed. Its likely possible to ship this (or a subset of this) extension so that we don't need to pull it out of `pip` or similar.

**Comment by reneme** [comment]
> I think there are some major unresolved issues here and since our adopting vs not adopting this is not a SemVer issue, I propose postponing to after 3.0 is released.

I've jotted down my thoughts on this in https://github.com/randombit/botan/issues/3377#issuecomment-1505155573. Let's move further discussion there.

TL;DR: In its current shape, Breathe will unfortunately not work for us, I'm afraid.

**Comment by reneme** [comment]
Rebased to latest master, added `sphinx-togglebutton` to the repo be independent of distro's repos and disabled `XML_PROGRAMLISTING`.

**Comment by randombit** [comment]
With #4185 merging the doc  comments, and Breathe apparently abandoned/on hold upstream I think we should close this and related PRs
